snuba/pipeline/query_pipeline.py

```diff
@@ -1,31 +1,41 @@
 from abc import ABC, abstractmethod
 
-
-from snuba.datasets.plans.query_plan import QueryRunner
+from snuba.datasets.plans.query_plan import ClickhouseQueryPlan, QueryRunner
+from snuba.query.logical import Query as LogicalQuery
 from snuba.request import Request
+from snuba.request.request_settings import RequestSettings
 from snuba.web import QueryResult
 
 
-class QueryPipeline(ABC):
+class EntityQueryProcessingPipeline(ABC):
+    """
+    A EntityQueryProcessingPipeline contains a series of steps that, given
+    a logical single entity query and request settings, performs the
+    query processing steps to the point where the query is ready to
+    be executed.
+
+    the query is returned as a ClickhouseQueryPlan.
     """
-    Contains the instructions to build the query plans for the requested query.
-    The QueryPipeline runs the query plan processors, executes the query plans and
-    returns the relevant result.
 
-    Most of the time, a single query plan is built by the SingleQueryPlanPipeline,
+    @abstractmethod
+    def execute(self) -> ClickhouseQueryPlan:
+        raise NotImplementedError
+
+
+class QueryExecutionPipeline(ABC):
```

Comment:
@lynnagara  what if we turned all these classes into functions. The pipeline builder can return a callable (which would be a partially applied function).
In the end we always need to build these objects on the fly whether it is for the multiple storage pipeline or for the subqueries in the composite query. We never define a pipeline statically.

Reply:
Turning this into a callable seems fine to me, it doesn't seem to do all that much except provide the execute method. Are you planning to keep an overall pipeline class though (ie. that represents the entire simple or composite pipeline), even if each of the segments of that pipeline are no longer pipelines themselves?

Reply:
I don't think a pipeline class will be needed. A pipeline can easily be either:
`Callable[[Query, RequestSettings], QueryResult]`
or just
`Callable[[], QueryResult]`
depending on how we provide the function as it can be a partially applied one.

I came to this conclusion because we are using classes as partially applied functions:
- They are strictly throw away object as they have to be built on the fly and the state is not valid for a second call
- Using callables captures this stateless nature and reduces the confusion on when to initialize the pipeline objects.